### DISCUSSION TITLE
(Java Agent)Stack tree completeness of performance profile?

### DISCUSSION THREAD

**Discussion by biuld** [apache/skywalking#11763] [Q&A] [answered]
Does skywalking provides any guarantee on the completeness of the stack tree generated by performance profile? Found this [article](https://skywalking.apache.org/docs/main/v9.1.0/en/guides/backend-profile/), but it doesn't mention the completeness. I want to utilize the profile ability to obtain all the function calls of a trace, but I have concerns about its completeness.

**Comment by biuld** [comment]
My concerns is because I've looked into the source code of the java agent. The `ProfileThread` picks up `StackTraceElement` at a constant pace(`threadDumpPeriod`) and has a `DUMP_MAX_STACK_DEPTH`. How can this strategy guarantees the completeness of the final merged stack tree?

**Reply by biuld** [reply]
Also, I doubt this method can generate a stable stack tree for multiple profile of the same endpoint.

**Comment by wu-sheng** [comment]
That threshold is protecting your system not crashed. As this APM is used for the product env.

> I want to utilize the profile ability to obtain all the function calls of a trace, but I have concerns about its completeness.

That should not be done in the product env. That doesn't make any sense. It is crazy use case.

**Reply by biuld** [reply]
Nope, not on production. My ultimate goal is to generate a runtime call graph for QA analysis purpose.

**Reply by wu-sheng** [reply, marked as answer]
It doesn't matter you are in product or not, SkyWalking is designed for product. So, we would not provide such feature.

**Reply by wu-sheng** [reply]
This means SkyWalking doesn't fit your requirement.

**Reply by biuld** [reply]
Noted, thanks for the confirmation.

**Reply by biuld** [reply]
This is not a feature request, actually I am writing my own plugin to achieve my goal. Just need some hard evidence on my assumptions.
